cpp/src/arrow/dataset/discovery.h

```diff
@@ -85,12 +79,20 @@ class ARROW_DS_EXPORT DataSourceDiscovery {
  protected:
   DataSourceDiscovery();
 
-  std::shared_ptr<Schema> schema_;
-  PartitionSchemePtr partition_scheme_;
   ExpressionPtr root_partition_;
 };
 
 struct FileSystemDiscoveryOptions {
+  // Either an explicit PartitionScheme or a PartitionSchemeDiscovery to discover one.
+  //
+  // If a discovery is provided, it will be used to infer a schema for partition fields
+  // based on file and directory paths then construct a PartitionScheme. The default is
+  // to discover a partition scheme assuming hive style directory naming.
+  //
+  // The resulting partition scheme will be applied to discovered files and the resulting
+  // partition information embedded in the DataSource.
+  PartitionSchemeOrDiscovery partition_scheme{HivePartitionScheme::MakeDiscovery()};
```

Comment:
This has the default of unpleasant surprises if a user doesn't expect it, the default should be none.

Reply:
I have no opinion on this; I was just following the behavior of `pyarrow.ParquetDataset`. I'll revert it but It seems like a reasonable default for the scripting languages at least (similar to inserting implicit casts). @nealrichardson @kszucs you may consider adding `HivePartitionScheme::MakeDiscovery()` as the default for this field in R/Python

Reply:
Python: batteries included. C++: sharp knives scattered on the floor.

Reply:
... and nothing is on by default